torch/onnx/symbolic_opset9.py

```diff
@@ -3886,7 +3886,7 @@ def new_zeros(
     g: jit_utils.GraphContext, self, sizes, dtype, layout, device, pin_memory=False
 ):
     self_dtype = symbolic_helper._try_get_scalar_type(self)
-    if dtype is None and self_dtype is not None:
+    if dtype.node().mustBeNone() and self_dtype is not None:
```

Comment:
This is using inter torchscript IR API, which is not ideal.

We have abstracted that away through [JitScalarType](https://github.com/pytorch/pytorch/blob/main/torch/onnx/_type_utils.py#L61) class that is called under the hood by `_try_get_scalar_type`

Why the current code is not good and how can we fix it without accessing private API from TorchScript here?

Reply:
I gave some detail about the issue with the current code here, https://github.com/pytorch/pytorch/issues/110597#issuecomment-1750331085, related to a (reason unidentified) change of behavior in torch.jit.trace.

I guess we could also modify https://github.com/pytorch/pytorch/blob/e9ebda29d87ce0916ab08c06ab26fd3766a870e5/torch/onnx/symbolic_helper.py#L78-L79 but it is unclear to me how so I modified directly the `new_zeros` function.

It seems that calling `_type_utils.JitScalarType.from_value(dtype)` on `(%15 : NoneType = prim::Constant(), scope: __main__.MyModel::)` raises:

```
torch.onnx.errors.SymbolicValueError: Cannot determine scalar type for this '<class 'torch.NoneType'>' instance and a default value was not provided.  [Caused by the value '15 defined in (%15 : NoneType = prim::Constant(), scope: __main__.MyModel::
)' (type 'NoneType') in the TorchScript graph. The containing node has kind 'prim::Constant'.] 
```

Reply:
Changing `_type_utils.JitScalarType.from_value(dtype)` would be a better solution. It is ok to use TorchScript API there because this is the abstraction layer between torchscript and onnx exporter

Reply:
Thank you. As you commented in the issue, it could be that the fix will be in torchscript anyway so I'll wait for an answer on that.

Reply:
dtype is an optional based on https://github.com/pytorch/pytorch/blob/80ea8784f392959f8d61e4cd37420819d5694bf1/aten/src/ATen/native/native_functions.yaml#L2343. So it is a valid condition we should account for imo.

Reply:
A regression has been acknowledged by PyTorch Core. By 10/13 we will have more information on how to address this upstream too